### ISSUE TITLE
Faulty geometry in IfcReinforcingBars

### ISSUE THREAD

**Issue by urosjovanovic** [xBimTeam/XbimGeometry#74]
Hello all,

I'm submitting a smaller, isolated sample where some of the IfcReinforcingBar geometry is not read correctly. Attached below are the sample file, output from Solibri Modelviewer, Xbim (current develop, the same is on master).

Here is the error list that was generated in Xbim:

```
==== 1:08 PM	DEBUG	Xbim.Geometry.Engine.Interop.XbimCustomAssemblyResolver
Loading assembly from: G:\---\Xbim\develop\XbimWindowsUI\Output\Debug\x86\Xbim.Geometry.Engine32.dll


==== 1:08 PM	DEBUG	Xbim.Geometry.Engine.Interop.XbimCustomAssemblyResolver
Loading assembly from: G:\---\Xbim\develop\XbimWindowsUI\Output\Debug\x86\Xbim.Geometry.Engine32.dll


==== 1:08 PM	ERROR	Xbim.Geometry.XbimGeometryCreator
GeomEngine: #11613=IfcSweptDiskSolid [Error creating geometry #11613 representation of type IfcSweptDiskSolid, BRep_API: command not done]


==== 1:08 PM	INFO	Xbim.ModelGeometry.Scene.Xbim3DModelContext
GeomScene: #11613=IfcSweptDiskSolid [Is an empty shape]


==== 1:08 PM	ERROR	Xbim.Geometry.XbimGeometryCreator
GeomEngine: #11744=IfcSweptDiskSolid [Error creating geometry #11744 representation of type IfcSweptDiskSolid, BRep_API: command not done]


==== 1:08 PM	INFO	Xbim.ModelGeometry.Scene.Xbim3DModelContext
GeomScene: #11744=IfcSweptDiskSolid [Is an empty shape]


==== 1:08 PM	ERROR	Xbim.Geometry.XbimGeometryCreator
GeomEngine: #11875=IfcSweptDiskSolid [Error creating geometry #11875 representation of type IfcSweptDiskSolid, BRep_API: command not done]


==== 1:08 PM	INFO	Xbim.ModelGeometry.Scene.Xbim3DModelContext
GeomScene: #11875=IfcSweptDiskSolid [Is an empty shape]


==== 1:08 PM	ERROR	Xbim.Geometry.XbimGeometryCreator
GeomEngine: #12006=IfcSweptDiskSolid [Error creating geometry #12006 representation of type IfcSweptDiskSolid, BRep_API: command not done]


==== 1:08 PM	INFO	Xbim.ModelGeometry.Scene.Xbim3DModelContext
GeomScene: #12006=IfcSweptDiskSolid [Is an empty shape]


==== 1:08 PM	ERROR	Xbim.Geometry.XbimGeometryCreator
GeomEngine: #11483=IfcSweptDiskSolid [Error creating geometry #11483 representation of type IfcSweptDiskSolid, BRep_API: command not done]


==== 1:08 PM	INFO	Xbim.ModelGeometry.Scene.Xbim3DModelContext
GeomScene: #11483=IfcSweptDiskSolid [Is an empty shape]


==== 1:08 PM	DEBUG	Xbim.Presentation.LayerStyling.SurfaceLayerStyler
Time to load visual components: 0.182 seconds
```

Looking forward to your answer.

Best regards,
Uros

[IfcReinforcingBar-GeometryIssue.zip](https://github.com/xBimTeam/XbimGeometry/files/1390799/IfcReinforcingBar-GeometryIssue.zip)
![rebar-solibri](https://user-images.githubusercontent.com/5221157/31662091-5e72d1ba-b33d-11e7-823f-a72d0054b434.png)
![rebar-xbim](https://user-images.githubusercontent.com/5221157/31662092-5e8a7dec-b33d-11e7-8f7d-3ee1ced48352.png)

**Comment by urosjovanovic** [comment]
Update:

Just tested with the latest develop version. It is still not loading correctly.
The visual output is still the same but no errors were generated.

**Comment by urosjovanovic** [comment]
Update 2:

I took a look at the code, there seems to be an issue with IfcSweptDiskSolid trimming. The problematic code lies [here](https://github.com/xBimTeam/XbimGeometry/blob/8ff50b4f013ca2da4291c3313277b2b1f4b7cea8/Xbim.Geometry.Engine/XbimSolid.cpp#L1293).

This line is especially interesting:
`// if the last parameter is about 1, use the lenght`
`double last  = Math::Abs(swdSolid->EndParam.Value - 1.0) < Precision::Confusion() ? sweep->Length : swdSolid->EndParam.Value;`

Does this mean it is assumed that StartParam and EndParam are in range [0, 1]? [The official docs for IfcSweptDiskSolid](http://www.buildingsmart-tech.org/ifc/IFC2x3/TC1/html/ifcgeometricmodelresource/lexical/ifcsweptdisksolid.htm) make no remarks regarding the value of these parameters. To make things even more confusing, Revit's exporter is filling these parameters with ANGLE values as you can see [here](https://sourceforge.net/p/ifcexporter/code/HEAD/tree/trunk/2017/Source/Revit.IFC.Export/Exporter/RebarExporter.cs). The line 361 says:
`endParam += UnitUtil.ScaleAngle(baseCurve.GetEndParameter(1) - baseCurve.GetEndParameter(0));`

So you get an entry like this:
`#6093= IFCSWEPTDISKSOLID(#6086,6.,$,0.,183.);`

The Xbim geometry generator will read 183 as length measured in default units (mm in this case) and it will chop the large portion of the rebar, so you end up with the output above.

My proposal is to ditch trimming all together and just use the entire curve, since it is not clear both by the standard and by authoring tools how the start/end parameters are used. Solibri seems to be doing exactly that, I've tried testing with different start/end param values and it always produces the same output.

This issue is actually related to one I've submitted earlier. #53

**Comment by CBenghi** [comment]
@urosjovanovic,
we have seen cases where that trimming is necessary to process the files though.
Particularly when curve then need to join up.
Feel free to adapt your own version and use that, but I'd have conflicting feelings about a PR on this.

I have no clear solution in mind at the moment, and I'm extra busy with lectures. It will probably be 2 months before I can go back to development with some serious time at hand.

**Comment by urosjovanovic** [comment]
@CBenghi 
That is fair enough, thanks for the prompt response.
To be clear, I totally understand that ditching the IfcSweptDiskSolid Start/End params is not the correct way to go implementation wise, but, as always, the end user just wants to see the correct _output_.

In the meantime, I'll try posting on Revit IFC forums to see if I can get some clarification for the calculation of endParam. I'll report back here.

**Comment by SteveLockley** [comment]
I have just reviewed this and found an issue with the tangential continuity of disks swept along the directrix. It should be noted that polylines should not exist in this representation unless they are G1 continuous. If they are not the IfcSweptDiskSolidPolygonal should be used.
This causes the sample ifc code above to crash when trying to construct a solid. I have fixed this so there will be a valid output, However, I am really confused as to what is supposed to happen here with the trim parameters. Typical only radius parametric values are used when the directrix is a conic. In this case they are not so the params should be length along the wire. I am taking the view that the other tools may show what you expect but they are not obeying the Ifc schema. We have always striven to make Xbim accurate and standards compliant, so I am not inclined to ignore these parameters. Its an error in the exporters. However, its a matter of belief so I have add an app.config parameter IgnoreIfcSweptDiskSolidParams if you set this to true it will ignore the params.
I am also minded to remove the cludge for exporters that use a parametric value between 0 and 1. I think the Ifc standard is now clear enough on this but that's for another day I guess.
These changes are in the develop branch pending release 5
